subprojects/language-java/src/integTest/groovy/org/gradle/language/java/JavaLanguageIncrementalBuildIntegrationTest.groovy

```diff
@@ -38,4 +38,33 @@ class JavaLanguageIncrementalBuildIntegrationTest extends AbstractJvmLanguageInc
         executedAndNotSkipped ":compileMainJarMainJava", ":createMainJar", ":mainJar"
     }
 
-}
\ No newline at end of file
+    def "task outcome is up to date when no recompilation necessary"() {
+        given:
+        buildFile << '''
+            tasks.withType(org.gradle.api.tasks.compile.AbstractCompile) { task ->
+                task.options.incremental = true
+                task.inputs.file 'input.txt' 
```

Comment:
You could also just add an unused dependency to the compilation classpath, I think that'll cause the task to run, but shouldn't require a recompilation?

Otherwise, a multi-project build makes the most sense.